File: frontend/src/app/domain/calc/recurrence.ts
import { ProjectedTransaction, RecurringFrequency, RecurringRule } from '../models/recurring';
import { addDays, addMonthsClamped, formatIsoDate, parseIsoDate } from './dates';

/**
 * The date of the Nth occurrence (`periods` = rule.interval × N), computed
 * directly from the rule's start date every time rather than by stepping
 * from the previous occurrence. That matters for month-end anchors: a
 * monthly rule starting Jan 31 must land on Mar 31 (not Mar 28) even
 * though the Feb occurrence in between clamped to 28 — stepping from a
 * clamped date would permanently lose the day-31 anchor, recomputing from
 * `start` every time doesn't.
 */
function occurrenceDate(start: Date, frequency: RecurringFrequency, periods: number): Date {
  switch (frequency) {
    case 'weekly':
      return addDays(start, 7 * periods);
    case 'monthly':
      return addMonthsClamped(start, periods);
    case 'yearly':
      return addMonthsClamped(start, 12 * periods);
  }
}

// A pathological rule (interval 0 or a huge range) shouldn't be able to
// spin this loop forever.
const MAX_OCCURRENCES = 2000;

/**
 * Expands a RecurringRule into its projected occurrences within
 * [rangeStart, rangeEnd] (inclusive, ISO dates). These are projections
 * only — see RecurringRule's doc comment — never persisted, and callers
 * must keep them out of balance/total/budget calculations.
 */
export function projectOccurrences(
  rule: RecurringRule,
  rangeStart: string,
  rangeEnd: string
): ProjectedTransaction[] {
  if (rule.interval < 1) return [];

  const start = parseIsoDate(rule.startDate);
  const rangeStartTime = parseIsoDate(rangeStart).getTime();
  const rangeEndTime = parseIsoDate(rangeEnd).getTime();
  const ruleEndTime = rule.endDate ? parseIsoDate(rule.endDate).getTime() : undefined;

  const occurrences: ProjectedTransaction[] = [];

  for (let index = 0; index < MAX_OCCURRENCES; index++) {
    const date = occurrenceDate(start, rule.frequency, rule.interval * index);
    const time = date.getTime();

    if (time > rangeEndTime) break;
    if (ruleEndTime !== undefined && time > ruleEndTime) break;

    if (time >= rangeStartTime) {
      occurrences.push({
        ...rule.template,
        date: formatIsoDate(date),
        recurringRuleId: rule.id,
        isProjected: true
      });
    }
  }

  return occurrences;
}

File: backend/app/services/recurrence.py
"""Occurrence-date math for a RecurringRule - the Python twin of the
frontend's domain/calc/recurrence.ts, kept in lockstep because both sides
must agree on which date a given occurrence falls on.

The Nth occurrence (`periods = rule.interval * N`) is computed directly
from the rule's start date every time, never by stepping from the previous
occurrence - that matters for month-end anchors: a monthly rule starting
Jan 31 must land on Mar 31 (not Mar 28) even though the Feb occurrence in
between clamped to 28. Stepping from a clamped date would permanently lose
the day-31 anchor; recomputing from `start` every time doesn't.
"""

import calendar
from datetime import date, timedelta

# A pathological rule (interval 0 or a huge range) shouldn't be able to
# spin this loop forever.
MAX_OCCURRENCES = 2000


def add_months_clamped(start: date, months: int) -> date:
    """Adds calendar months, clamping the day-of-month to whatever the
    target month actually has - Jan 31 + 1 month lands on Feb 28 (or 29),
    not an overflowed Mar 3. `months` may be negative."""
    total = start.month - 1 + months
    year = start.year + total // 12
    month = total % 12 + 1
    last_day = calendar.monthrange(year, month)[1]
    return date(year, month, min(start.day, last_day))


def occurrence_date(start: date, frequency: str, periods: int) -> date:
    if frequency == "weekly":
        return start + timedelta(days=7 * periods)
    if frequency == "monthly":
        return add_months_clamped(start, periods)
    if frequency == "yearly":
        return add_months_clamped(start, 12 * periods)
    raise ValueError(f"unknown frequency: {frequency}")


def project_occurrence_dates(
    *,
    start_date: date,
    frequency: str,
    interval: int,
    end_date: date | None,
    range_start: date,
    range_end: date,
) -> list[date]:
    """Every occurrence date within [range_start, range_end] (inclusive),
    truncated by the rule's own end_date. Mirrors projectOccurrences in
    recurrence.ts, minus the template payload - callers here only need the
    dates."""
    if interval < 1:
        return []

    dates: list[date] = []
    for index in range(MAX_OCCURRENCES):
        occurrence = occurrence_date(start_date, frequency, interval * index)
        if occurrence > range_end:
            break
        if end_date is not None and occurrence > end_date:
            break
        if occurrence >= range_start:
            dates.append(occurrence)

    return dates
